vhost_user_fs/src/passthrough.rs

```diff
@@ -911,6 +913,62 @@ impl FileSystem for PassthroughFs {
         self.do_unlink(parent, name, 0)
     }
 
+    fn setupmapping<T: FsCacheReqHandler>(
+        &self,
+        _ctx: Context,
+        inode: Inode,
+        _handle: Handle,
+        foffset: u64,
+        len: u64,
+        flags: u64,
+        moffset: u64,
+        vu_req: &mut T,
+    ) -> io::Result<()> {
+        debug!(
+            "setupmapping: ino {:?} foffset {} len {} flags {} moffset {}",
+            inode, foffset, len, flags, moffset
+        );
+
+        let mut msg: VhostUserFSSlaveMsg = Default::default();
+        msg.fd_offset[0] = foffset;
+        msg.cache_offset[0] = moffset;
+        msg.len[0] = len;
+        msg.flags[0] = if (flags & fuse::SetupmappingFlags::WRITE.bits()) != 0 {
+            VhostUserFSSlaveMsgFlags::MAP_W | VhostUserFSSlaveMsgFlags::MAP_R
+        } else {
+            VhostUserFSSlaveMsgFlags::MAP_R
+        };
+
+        let open_flags = if (flags & fuse::SetupmappingFlags::WRITE.bits()) != 0 {
+            libc::O_RDWR
+        } else {
+            libc::O_RDONLY
+        };
+
+        let file = self.open_inode(inode, open_flags as i32)?;
+        (*vu_req)
+            .map(&msg, file.as_raw_fd())
+            .or_else(|_| Err(io::Error::from_raw_os_error(libc::EINVAL)))
+    }
+
+    fn removemapping<T: FsCacheReqHandler>(
+        &self,
+        _ctx: Context,
+        requests: Vec<fuse::RemovemappingOne>,
+        vu_req: &mut T,
+    ) -> io::Result<()> {
+        let mut msg: VhostUserFSSlaveMsg = Default::default();
+
+        for (ind, req) in requests.iter().enumerate() {
+            msg.len[ind] = req.len;
```

Comment:
Should probably check to make sure that `requests.len() < VHOST_USER_FS_SLAVE_ENTRIES`.  Otherwise this will panic and abort the whole process.

Reply:
Thanks for pointing this out, it does cause problem because requests may have more items than `VHOST_USER_FS_SLAVE_ENTRIES`, I'm refactoring the code a little bit.